sdk/program/src/lib.rs

```diff
@@ -1,3 +1,650 @@
+//! that will fail in off-chain scenarios at runtime. This distinction is not
+//! well-reflected in the documentation.
+//!
+//! For a more complete description of Solana's implementation of eBPF and its
+//! limitations, see the main Solana documentation on [on-chain programs][ocp].
+//!
+//! [ocp]: https://docs.solana.com/developing/on-chain-programs/overview
+//!
+//! # Core data types
+//!
+//! - [`Pubkey`] - An [ed25519] public key. All Solana accounts are identified
+//!   by a `Pubkey`, though some may not have corresponding private keys. As
+//!   Solana programs do not handle secret keys, the full [`Keypair`] is not
+//!   defined in `solana-program` but in `solana-sdk`.
+//! - [`Hash`] - A [SHA-256] hash, used to uniquely identify blocks.
+//! - [`AccountInfo`] - A description of a single Solana account. All accounts
+//!   that might be accessed by a program invocation are provided to the program
+//!   entrypoint as `AccountInfo`.
+//! - [`Instruction`] - A directive telling the runtime to execute a program,
+//!   passing it a set of accounts and program-specific data.
+//! - [`ProgramError`] and [`ProgramResult`] - The error type that all programs
+//!   must return, reported to the runtime as a `u64`.
+//!
+//! [`Pubkey`]: pubkey::Pubkey
+//! [`Hash`]: hash::Hash
+//! [`Instruction`]: instruction::Instruction
+//! [`AccountInfo`]: account_info::AccountInfo
+//! [`ProgramError`]: program_error::ProgramError
+//! [`ProgramResult`]: entrypoint::ProgramResult
+//!
+//! [ed25519]: https://ed25519.cr.yp.to/
+//! [`Keypair`]: https://docs.rs/solana-sdk/latest/solana_sdk/signer/keypair/struct.Keypair.html
+//! [SHA-256]: https://en.wikipedia.org/wiki/SHA-2
+//!
+//! # Serialization
+//!
+//! Within the Solana runtime, programs, and network, at least three different
+//! serialization formats are used, and `solana-program` provides access to
+//! those needed by programs.
+//!
+//! In user-written Solana program code, serialization is primarily used for accessing
+//! [`AccountInfo`] data and [`Instruction`] data, both of which are program-specific
+//! binary data, the format up to the author.
+//!
+//! [`AccountInfo`]: account_info::AccountInfo
+//! [`Instruction`]: instruction::Instruction
+//!
+//! The three serialization formats in use in Solana are:
+//!
+//! - __[Borsh]__, a compact and well-specified format developed by the [NEAR]
+//!   project, suitable for use in protocol definitions and for archival storage.
+//!   It has a [Rust implementation][brust] and a [JavaScript implementation][bjs]
+//!   and is recommended for all purposes.
+//!
+//!   Users need to import the [`borsh`] crate themselves &mdash; it is not
+//!   re-exported by `solana-program`, though this crate provides several useful
+//!   utilities in its [`borsh` module][borshmod] that are not available in the
+//!   `borsh` library.
+//!
+//!   The `Instruction` type has built-in support for borsh.
```

Comment:
What is this referring to?

Reply:
The `new_with_borsh` constructor. I'll clarify.